src/natcap/invest/stormwater.py

```diff
@@ -0,0 +1,1332 @@
+
+    # (Optional) Aggregate to watersheds if an aggregate vector is defined
+    if (args['aggregate_areas_path']):
+        reproject_aoi_task = task_graph.add_task(
+            func=pygeoprocessing.reproject_vector,
+            args=(
+                args['aggregate_areas_path'],
+                target_raster_info['projection_wkt'],
+                FILES['reprojected_aoi_path']),
+            kwargs={'driver_name': 'GPKG'},
+            target_path_list=[FILES['reprojected_aoi_path']],
+            task_name='reproject aggregate areas vector to match rasters',
+            dependent_task_list=[]
+        )
+        aggregation_dependencies.append(reproject_aoi_task)
+        aggregation_task = task_graph.add_task(
+            func=aggregate_results,
+            args=(
+                FILES['reprojected_aoi_path'],
+                data_to_aggregate,
+                FILES['aggregate_data_path']),
+            target_path_list=[FILES['aggregate_data_path']],
+            dependent_task_list=aggregation_dependencies,
+            task_name='aggregate data over polygons'
+        )
+
+    task_graph.close()
+    task_graph.join()
+
+def threshold_array(array, threshold):
+    """Return a boolean array where 1 means less than the threshold value.
+    Assumes that the array's nodata value is the global NODATA."""
+    out = numpy.full(array.shape, NODATA, dtype=numpy.int8)
+    valid_mask = array != NODATA
+    out[valid_mask] = array[valid_mask] <= threshold
+    return out
+
+def ratio_op(lulc_array, soil_group_array, ratio_lookup, sorted_lucodes):
+    """Make an array of stormwater retention or infiltration ratios from 
+    arrays of LULC codes and hydrologic soil groups.
+
+    Args:
+        lulc_array (numpy.ndarray): 2D array of LULC codes
+        soil_group_array (numpy.ndarray): 2D array with the same shape as
+            ``lulc_array``. Values in {1, 2, 3, 4} corresponding to soil 
+            groups A, B, C, and D.
+        ratio_lookup (numpy.ndarray): 2D array where rows correspond to 
+            sorted LULC codes and columns correspond to soil groups
+            A, B, C, D in order. Shape: (number of lulc codes, 4)
+        sorted_lucodes (list[int]): List of LULC codes sorted from smallest 
+            to largest. These correspond to the rows of ``ratio_lookup``.
+
+    Returns:
+        2D numpy array with the same shape as ``lulc_array`` and 
+        ``soil_group_array``. Each value is the corresponding ratio for that
+        LULC code x soil group pair.
+    """
+    sorted_soil_groups = [1, 2, 3, 4]
+    # the index of each soil group in the sorted soil groups array
+    soil_group_index = numpy.digitize(soil_group_array, sorted_soil_groups, 
+        right=True)
+    # the index of each lucode in the sorted lucodes array
+    lulc_index = numpy.digitize(lulc_array, sorted_lucodes, right=True)
+    output_ratio_array = ratio_lookup[lulc_index, soil_group_index]
+    return output_ratio_array
```

Comment:
See explanation on the comment on the `lulc_array` for why I did this.
I still think it's correct for the `lulc_array`, but you're right it's not necessary here, because the soil groups are consecutive integers.

Reply:
I did forget to handle nodata values here (just fixed that). All the other functions do handle nodata values, and they use the global `NODATA` as much as possible, though that's explicitly documented. I made a related suggestion for pygeoprocessing [here](https://github.com/natcap/pygeoprocessing/issues/176).